dwallet/src/eth_state.rs

```diff
@@ -0,0 +1,720 @@
+                    }
+                }
+
+                if self.finalized_header.slot > self.optimistic_header.slot {
+                    self.optimistic_header = self.finalized_header.clone();
+                }
+            }
+
+            let finalized_header_checkpoint = self
+                .finalized_header
+                .hash_tree_root()
+                .map_err(|_| anyhow!("could not hash finalized header"))
+                .unwrap()
+                .encode_hex_with_prefix();
+
+            self.last_checkpoint = finalized_header_checkpoint;
+        }
+    }
+
+    fn verify_update(&self, update: &Update) -> Result<(), Error> {
+        let update = GenericUpdate::from(update);
+        self.verify_generic_update(&update)
+    }
+
+    fn verify_finality_update(&self, update: &FinalityUpdate) -> Result<(), Error> {
+        let update = GenericUpdate::from(update);
+        self.verify_generic_update(&update)
+    }
+
+    fn verify_optimistic_update(&self, update: &OptimisticUpdate) -> Result<(), Error> {
+        let update = GenericUpdate::from(update);
+        self.verify_generic_update(&update)
+    }
+
+    /// Verifies the correctness of a generic update received by the consensus client.
+    /// Validates a `GenericUpdate` based on several criteria to ensure it can be safely applied to
+    /// the client's state. The verification process includes checks for sufficient
+    /// participation, timing and period validity, relevance of the update, and the authenticity
+    /// of signatures.
+    ///
+    /// # Verification Process
+    /// 1. **Participation Check:** Verifies that the update has sufficient participation from the
+    ///    sync committee by checking the number of bits set in
+    ///    `sync_aggregate.sync_committee_bits`.
+    /// 2. **Timing Validation:** Ensures the update's timing is valid by comparing the
+    ///    `signature_slot` with the `attested_header.slot`, and the `finalized_header.slot`. The
+    ///    update must be signed after the attested header's slot and before or at the current slot,
+    ///    and it must reference a slot that is not older than the last finalized slot.
+    /// 3. **Period Validation:** Confirms that the update's signature slot falls within the correct
+    ///    sync committee period, allowing for updates from the current or immediately next period
+    ///    if the next sync committee is known.
+    /// 4. **Relevance Check:** Ensures the update is relevant by disallowing updates that reference
+    ///    slots older than the last finalized slot unless introducing a new sync committee.
+    /// 5. **Finality and Next Sync Committee Proofs:** Validates proofs related to the update's
+    ///    finality and the inclusion of a new sync committee, if present.
+    /// 6. **Signature Verification:** Confirms that the sync committee's signature on the attested
+    ///    header is valid, indicating agreement with the header's contents.
```

Comment:
```suggestion
    /// Verifies the correctness of a generic update received by the consensus client.
    /// Validates a `GenericUpdate` based on several criteria to ensure it can be safely applied to
    /// the client's state. The verification process includes checks for sufficient
    /// participation, timing and period validity, relevance of the update, and the authenticity
    /// of signatures.
    ///
    /// # Verification Process
    /// 1. **Participation Check** – Verifies that the update has sufficient participation from the
    ///    sync committee by checking the number of bits set in
    ///    `sync_aggregate.sync_committee_bits`.
    /// 2. **Timing Validation** – Ensures the update's timing is valid by comparing the
    ///    `signature_slot` with the `attested_header.slot`, and the `finalized_header.slot`. The
    ///    update must be signed after the attested header's slot and before or at the current slot,
    ///    and it must reference a slot that is not older than the last finalized slot.
    /// 3. **Period Validation** Confirms that the update's signature slot falls within the correct
    ///    sync committee period, allowing for updates from the current or immediate next period if
    ///    the next sync committee is known.
    /// 4. **Relevance Check** – Ensures the update is relevant by disallowing updates that
    ///    reference slots older than the last finalized slot unless introducing a new sync
    ///    committee.
    /// 5. **Finality and Next Sync Committee Proofs** – Validates proofs related to the update's
    ///    finality and the inclusion of a new sync committee, if present.
    /// 6. **Signature Verification** – Confirms that the sync committee's signature on the attested
    ///    header is valid, indicating agreement with the header's contents.
```